### PR TITLE
Pull in CMSIS-DAP support from upstream Zephyr

### PR DESCRIPTION
**Pull request by maxd-nordic**

### CODE REVIEW

File: drivers/dp/swdp_ll_pin.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+#warning "Pin delay is not defined"
+#endif
+}
+
+static ALWAYS_INLINE void swdp_ll_pin_input(void *const base, uint8_t pin)
+{
+#if defined(CONFIG_SOC_SERIES_NRF52X) || defined(CONFIG_SOC_SERIES_NRF53X)
+	NRF_GPIO_Type * reg = base;
+
+	reg->PIN_CNF[pin] = 0b0000;
```
Comment (masz-nordic) [inline drivers/dp/swdp_ll_pin.h:43]: HAL should be used here, and in code below.

**Review by masz-nordic — CHANGES_REQUESTED** [review]

File: drivers/dp/swdp_ll_pin.h (hunk 1 above)
Comment (maxd-nordic) [reply drivers/dp/swdp_ll_pin.h:43]: this is from upstream zephyr - it's not the right place to change it here.

Comment (masz-nordic) [reply drivers/dp/swdp_ll_pin.h:43]: It should be fixed in a follow-up PR in upstream, that would be cherry-picked as fromlist to this PR.

Comment (maxd-nordic) [reply drivers/dp/swdp_ll_pin.h:43]: @masz-nordic Could you give me some pointers on how to do that? `nrf_gpio_cfg` uses quite a few more cycles then just setting the PIN_CNF register directly. Speed is of utmost importance for this part.

Comment (masz-nordic) [reply drivers/dp/swdp_ll_pin.h:43]: Could you post disassembly here? Is `nrf_gpio_pin_port_decode` not getting optimized enough?
We can add new functions if needed.

File: drivers/dp/swdp_ll_pin.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+#endif
+}
+
+
+static ALWAYS_INLINE void swdp_ll_pin_set(void *const base, uint8_t pin)
+{
+#if defined(CONFIG_SOC_SERIES_NRF52X) || defined(CONFIG_SOC_SERIES_NRF53X)
+	NRF_GPIO_Type * reg = base;
+
+	reg->OUTSET = BIT(pin);
```
Comment (masz-nordic) [inline drivers/dp/swdp_ll_pin.h:62]: Can be replaced with `nrf_gpio_port_out_set`

File: drivers/dp/swdp_ll_pin.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+#endif
+}
+
+static ALWAYS_INLINE void swdp_ll_pin_clr(void *const base, uint8_t pin)
+{
+#if defined(CONFIG_SOC_SERIES_NRF52X) || defined(CONFIG_SOC_SERIES_NRF53X)
+	NRF_GPIO_Type * reg = base;
+
+	reg->OUTCLR = BIT(pin);
```
Comment (masz-nordic) [inline drivers/dp/swdp_ll_pin.h:71]: Can be replaced with `nrf_gpio_port_out_clear`

File: drivers/dp/swdp_ll_pin.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+#endif
+}
+
+static ALWAYS_INLINE uint32_t swdp_ll_pin_get(void *const base, uint8_t pin)
+{
+#if defined(CONFIG_SOC_SERIES_NRF52X) || defined(CONFIG_SOC_SERIES_NRF53X)
+	NRF_GPIO_Type * reg = base;
+
+	return ((reg->IN >> pin) & 1);
```
Comment (masz-nordic) [inline drivers/dp/swdp_ll_pin.h:80]: `nrf_gpio_pin_read`

**Comment by tmon-nordic** [comment]
Is this going to be obsolete by upmerge?

**Comment by maxd-nordic** [comment]
We have a version of this as a patch on hello.nrfcloud/firmware. Since this is upstream, it will reach us eventually.

**Pull request closed without merging**